Auth Matrix app

sam-bertin opened the pull request:
https://taiga.maykinmedia.nl/project/maykin-intranet/issue/399 => Taiga Ticket

_The wish is (via the admin -> users) to be able to click through to a custom admin-view showing a matrix of authorizations for a project.

This matrix would be 2 tables: one showing the user-group relationships and one showing the group-permissions relationships. The groups would be at the top for both matrixes, and for every relationship we show a 'x' with the respective user or permission.

This should from this page also be exportable as an excel. If the admin-user has view-permissions for the users they should be able to see this admin-view and be able to export the authorization matrix._

sam-bertin committed:
Auth Matrix app

sam-bertin committed:
Update CI

sam-bertin committed:
Revert "Update CI"

This reverts commit e6012fe8a9da9df84ac68889e2162d8fb00538aa.

sam-bertin committed:
CI Update

sam-bertin committed:
Organize Imports

sam-bertin committed:
Black / Flake8

sam-bertin committed:
Sphinx Docs

sam-bertin committed:
Update CI

sam-bertin committed:
Match Project Name

sam-bertin committed:
Revert "Match Project Name"

This reverts commit 70e6a7d154d583ef6db0b6f608c4cb9f3b126059.

sam-bertin committed:
Publishing configurations

sam-bertin committed:
Activate Docs LinkCheck

sam-bertin committed:
FIX: CodeCov / ReadTheDocs links

CharString committed:
Remove token auth from Codecov step in  ci.yml

sam-bertin committed:
CI CodeConv Token

sam-bertin reviewed (commented):
As we saw yesterday it is in fact required to pass in the token.
We were on the right track as you figured out the way to find the token on the codecov website.
Then it needs to be set up in github in the repository settings (where we did not have access).

sam-bertin committed:
Project Version to publish to Pypi

sam-bertin committed:
Add LinkCheck

sam-bertin committed:
Update Documentation

sergei-maertens commented on .github/workflows/documentation-links.yml:
Hrm, I remember just enabling the docs build (for PRs too) in the RTD configuration interface, this action is not something we have in other apps. Is it really needed?

sergei-maertens commented on .github/workflows/documentation-links.yml:
```diff
@@ -0,0 +1,22 @@
+    # Execute this action only on PRs that touch
+    # documentation files.
+    paths:
+      - "docs/**"
```
when using `sphinx.ext.autodoc`, this would not rebuild docs on source code changes, while it should. I think I'd leave this unspecified.

sergei-maertens commented on CHANGELOG.rst:
Can you stick to the header styles used in other libraries, e.g. https://github.com/maykinmedia/django-admin-index/blob/master/CHANGELOG.rst?plain=1

sergei-maertens commented on README.rst:
```diff
@@ -1,18 +1,18 @@
 
-Welcome to auth_matrix's documentation!
+Welcome to maykin-auth-matrix's documentation!
 =================================================
```
line lengths don't match up

sergei-maertens commented on pyproject.toml:
```diff
@@ -3,8 +3,8 @@ requires = ["setuptools>=61.0.0"]
-name = "auth_matrix"
-version = "0.1.0"
+name = "maykin-auth-matrix"
+version = "0.1.1"
```
it looks like you're manually updating the version numbers, since I don't see the `current_version` updated in the `[tool.bumpversion]` section.

Please check the documentation for bump-my-version to see how you should manage the version numbers.

sam-bertin replied:
let's bump it after review if I understood correctly (review->merge->bump->tag->release)

sergei-maertens commented on README.rst:
```diff
@@ -1,18 +1,18 @@
-:Source: https://github.com/maykinmedia/auth_matrix
+:Version: 0.1.1
+:Source: https://github.com/maykinmedia/maykin-auth-matrix
 :Keywords: ``<keywords>``
```
please fill this out

sergei-maertens commented on README.rst:
```diff
@@ -1,18 +1,18 @@
 |python-versions| |django-versions| |pypi-version|
 
-<One liner describing the project>
+DigiD Compliant Authorization Matrix for the administation of Django projects.
```
```suggestion
Generate and export an authorization matrix of Django (staff) users.

DigiD audit requirements require insight in user authorizations, provided through these matrices.
```

sergei-maertens commented on README.rst:
```diff
@@ -21,74 +21,198 @@ Welcome to auth_matrix's documentation!
-* ...
-* ...
+* Display a matrix of permissions for the groups
+* Display a matrix of groups for the users
```
maybe replace `Display` with `Generate`?

sergei-maertens commented on README.rst:
```diff
@@ -21,74 +21,198 @@ Welcome to auth_matrix's documentation!
+To use this with your project you need to follow these steps:
+
+#. Add ``maykin-auth-matrix`` to ``INSTALLED_APPS`` in
+   your Django project's ``settings.py``.:
```
Let's just point to the documentation on RTD here in the readme, and move these instructions into `docs/quickstart.rst`.

sergei-maertens commented on README.rst:
```diff
@@ -21,74 +21,198 @@ Welcome to auth_matrix's documentation!
+          "maykin-auth-matrix",
+      )
+
+   Note that there is no dash in the module name, only underscores.
```
installed apps entry should be `auth_matrix`, it's the python path that gets installed, not the package name.

The note can be removed, dashes are not valid characters in python identifiers anyway.

sergei-maertens commented on README.rst:
```diff
@@ -21,74 +21,198 @@ Welcome to auth_matrix's documentation!
+   .. code-block:: console
+
+      $ python manage.py migrate maykin-auth-matrix
+
```
Can you follow the same structure as https://github.com/maykinmedia/django-simple-certmanager/blob/main/docs/quickstart.rst?plain=1?

Most notably, the app name should not be specified as a `migrate` argument, this can mess with the migration dependencies. Just `python manage.py migrate` is enough.

sergei-maertens commented on README.rst:
```diff
@@ -21,74 +21,198 @@ Welcome to auth_matrix's documentation!
+
+    from django.contrib import admin
+    from django.urls import path, include
+    from auth_matrix.views import AuthMatrixView
```
```suggestion
    from django.contrib import admin
    from django.urls import path, include

    from auth_matrix.views import AuthMatrixView
```
different import groups according to isort config, so let's give the right examples

sergei-maertens commented on README.rst:
```diff
@@ -21,74 +21,198 @@ Welcome to auth_matrix's documentation!
+
+    urlpatterns = [
+        path(
+            "admin/authorization/",
```
let's set up examples with URL namespaces equal to the package name:

```suggestion
            "admin/auth_matrix/",
```

sergei-maertens commented on auth_matrix/admin.py:
```diff
@@ -40,15 +43,22 @@ def dehydrate(self, user):
         return data
 
     def export(self, queryset, *args, **kwargs):
+        queryset = self.get_queryset()
```
why does this override the `queryset` argument from the method itself?

sergei-maertens commented on auth_matrix/admin.py:
```diff
@@ -40,15 +43,22 @@ def dehydrate(self, user):
+        queryset = self.get_queryset()
         dataset = super().export(queryset, *args, **kwargs)
         group_names = [group.name for group in Group.objects.all()]
         for group_name in group_names:
             dataset.append_col(
-                [user.groups.filter(name=group_name).exists() for user in queryset],
+                [
+                    user.groups.filter(name=group_name).exists()
+                    for user in User.objects.all()
+                ],
                 header=group_name,
             )
```
this is a `n+1` query - 1 query to retrieve all the groups, and then for each group a query is done to retrieve the users. The more groups you have, the worse performance gets.

Instead, you can annotate each user to get a list of its group names with each user object.

```py
group_names_subquery = (
    Group.objects
    .filter(user=OuterRef('pk'))
    .values('user')
    .annotate(group_names=ArrayAgg('name'))
    .values('group_names')
)
queryset = queryset.annotate(group_names=Subquery(group_names_subquery))
```
This results in 1 query to retrieve all the users with the group names attached to it.

Next, you can (efficiently) loop over the group names to add the extra columns (that will be query 2):

```py
for group_name in Group.objects.values_list("name", flat=True).order_by("name"):
    column = [group_name in user.group_names for user in queryset]
    ...
```
make sure to loop over `queryset` again since it will re-use the queryset cache and not create additional queries (and to guarantee the ordering, see paragraph below).

Note that you must make sure to iterate over the same `queryset` to make sure the order of rows added in the extra column is stable! If you sort the admin on `-username` for example, and `User.objects.all()` ordering is undefined (let's assume it orders on PK then, but it might be random), then your matrix will be wrong. I'd like to see a test for this!

sam-bertin replied:
```python
    def export(self, queryset, *args, **kwargs):
        group_names_subquery = (
            Group.objects.filter(user=OuterRef("pk"))
            .values("user")
            .annotate(group_names=ArrayAgg("name"))
            .values("group_names")
        )
        queryset = super().get_queryset()  # Else the exported CSV user values are blank
        queryset = queryset.annotate(group_names=Subquery(group_names_subquery))

        dataset = super().export(queryset, *args, **kwargs)

        for group_name in Group.objects.values_list("name", flat=True).order_by("name"):
            column = [group_name in (user.group_names or []) for user in queryset]
            dataset.append_col(column, header=group_name)

        return dataset
```
The user queryset is just a single blank line if I do not use `queryset = super().get_queryset()` forcing it is the only way I found to make it work

sergei-maertens commented on auth_matrix/admin.py:
```diff
@@ -40,15 +43,22 @@ def dehydrate(self, user):
             )
         return dataset
 
+    def get_queryset(self):
+        return super().get_queryset()
```
this doesn't seem to do anything :-)

sergei-maertens commented on auth_matrix/admin_views.py:
```diff
@@ -6,6 +6,7 @@
 User = get_user_model()
 
+
 @method_decorator(staff_member_required, name="dispatch")
```
we tend to follow a stricter approach to add proper permission checks - staff user does not automatically mean you should get access to this information! The ticket also mentions that it should be connected to the permission to view users.

You can look at https://github.com/open-formulieren/open-forms/blob/master/src/openforms/config/admin_views.py#L11 for inspiration and you typically also want to wrap this properly with `admin.site.admin_view`: https://github.com/open-formulieren/open-forms/blob/b0feb8aa60d10c16ede0873420570a5d4668d3e2/src/openforms/admin/urls.py#L31

sam-bertin replied:
Yeah ok. I have tried to account for both definition of the User but could only test with auth.User model.

sergei-maertens commented on auth_matrix/templates/admin/auth/user/change_list.html:
This will not work properly in projects where a custom user model is used OR a custom template is already used.

IMO this belongs in the extra actions at the top, just like in intranet for hour registration.

I would provide the minimal stuff as a template that can be included like `{% include "auth_matrix/includes/user_actions.html" %}` so that downstream projects have flexibility (and the documentation should then mention how to use this).

sam-bertin replied:
Yeah if the user is not defined from auth/accounts then the Show Authorization Matrix button won't show as only this admin template is being overridden. That is why we have agreed with Alex to only have the button mandatory for Groups and optional for Users. auth.user and account.user will be showing the button but that is not required.

sergei-maertens commented on auth_matrix/templates/admin/auth/user/change_list.html:
```diff
@@ -0,0 +1,27 @@
+
+{% block extrastyle %}
+{{ block.super }}
+<style>
+    .extra {
+        padding: 1em;
+    }
+
+    .is-flex {
+        display: flex;
+    }
+</style>
```
inline `<style>` is blegh, if you use the proper classes for the user action buttons that the django admin ships with, you should not need any additional CSS

sergei-maertens commented on auth_matrix/templates/admin/auth/user/change_list.html:
```diff
@@ -0,0 +1,27 @@
+
+{% block pagination %}
+    {{ block.super }}
+    {% if request.user.is_superuser %}
```
this does not match the permissions defined on the view itself, which currently only requires staff user permissions

Instead, this should use the proper admin view permission check, see https://docs.djangoproject.com/en/4.2/topics/auth/default/#permissions

sam-bertin replied:
it should be now

sergei-maertens commented on auth_matrix/templates/admin/auth/user/change_list.html:
```diff
@@ -0,0 +1,27 @@
+    {{ block.super }}
+    {% if request.user.is_superuser %}
+    <div class='extra is-flex'>
+        <h3>Extra Actions :</h3>
```
All user-facing strings/literals should be marked as translatable.

sergei-maertens commented on auth_matrix/templates/admin/auth/user/change_list.html:
```diff
@@ -0,0 +1,27 @@
+    {% if request.user.is_superuser %}
+    <div class='extra is-flex'>
+        <h3>Extra Actions :</h3>
+        <a href="{% url 'authorization_matrix' %}" class="button mt">
+            Show  the authorization matrix
+        </a>
```
please properly namespace the URL names, this should read `{% url "auth_matrix:view_matrix" %}` or something like that

sergei-maertens commented on docs/contributing.rst:
```diff
@@ -0,0 +1,85 @@
+Contributing
+============
+
+We have a strong vision about the scope of this package and the "less is more" principle
```
this is duplicated in another part of the docs, I would keep this section and remove the other place (README I think?)

sergei-maertens commented on docs/images/authorization_matrix.png:
Can you use the class names/markup like the django list view uses, so we get the same visual style for all tables in the admin?

sam-bertin replied:
I was not able to figure out useful classes or properties to match the admin tables styling.

sergei-maertens commented on docs/index.rst:
same remarks as in the README.rst apply here

sergei-maertens commented on docs/quickstart.rst:
```diff
@@ -9,10 +9,80 @@ Install from PyPI with pip:
+Requirements
+============
+
+This application makes use of the AbstractBaseUser model from django.
+Generally, you should be able to use this application with any Django project that uses the default User model.
+If you are using a custom User model, you will need to make sure that it is compatible with the AbstractBaseUser model.
+We made it compatible for the username and email fields, but you might need to make some changes to the code to make it work with your custom User model.
```
please avoid line wrapping an limit line lengths to 80 chars

sergei-maertens replied:
this should mention that things will work out of the box if you inherit from `AbstractUser`, and otherwise you can customize things through the `username` and `email` properties on your custom user model.

sergei-maertens commented on docs/quickstart.rst:
the remarks from the README apply here

sergei-maertens commented on py.test:
this file should not be here

sergei-maertens commented on python:
file should not be here

sergei-maertens commented on testapp/auth_matrix.db:
file should be excluded from version control

sergei-maertens commented on testapp/settings.py:
```diff
@@ -51,3 +53,20 @@
 
 ROOT_URLCONF = "testapp.urls"
+
+LOGGING = {
```
do we need this?

sergei-maertens commented on tests/test_admin_views.py:
please write the tests in pytest style, so no inheriting from `django.test.*`

You can take a look at https://github.com/maykinmedia/django-simple-certmanager/blob/main/tests/test_admin.py for inspiration

sam-bertin committed:
📝 Review pt.1

sergei-maertens commented on .gitignore:
```diff
@@ -46,6 +46,8 @@ coverage.xml
 *.cover
 .hypothesis/
 .pytest_cache/
+pytest.ini
```
this is a config file that should typically NOT be ignored - why was it added?

sergei-maertens commented on .gitignore:
```diff
@@ -102,3 +105,5 @@ venv.bak/
 # mypy
 .mypy_cache/
+
+testapp/auth_matrix.db
```
it's already listed on line 60

sergei-maertens commented on auth_matrix/admin.py:
```diff
@@ -40,13 +45,21 @@ def dehydrate(self, user):
         return data
 
     def export(self, queryset, *args, **kwargs):
```
I looked at the docs and `queryset` can be None`, which is a bit odd, but either way it doesn't seem right to override `export` itself when there are `before_export` and `after_export` hooks provided by the library: https://django-import-export.readthedocs.io/en/latest/api_resources.html#import_export.resources.Resource.after_export

Instead, you should be using the `after_export` hook here, you receive the (filtered) queryset and tablib dataset as arguments so you have all the information you need.

You can override `get_queryset` to add the `group_names` annotation

sam-bertin replied:
I have an issue implementing like this :

```python
    def get_queryset(self):
        group_names_subquery = (
            Group.objects.filter(user=OuterRef("pk"))
            .values("user")
            .annotate(group_names=ArrayAgg("name"), default=Value([]))
            .values("group_names")
        )
        users = super().get_queryset()
        users_with_group_names = users.annotate(
            group_names=Subquery(group_names_subquery)
        )
        return users_with_group_names

    def after_export(self, queryset, data, *args, **kwargs):
        # print of the query sets
        u = self.get_queryset()
        print(u.values())
        print(queryset.values())
        dataset = data
        for group_name in Group.objects.values_list("name", flat=True).order_by("name"):
            column = [group_name in (user.group_names or []) for user in queryset]
            dataset.append_col(column, header=group_name)
        breakpoint()
        return dataset
```

Here is the printed results : 

```
<QuerySet [{'id': 2, 'password': 'pbkdf2_sha256$600000$uV1L4D8LZriX6MygnONdoW$1R4sCpfGaFdqaIxPTwI4YxvZplsrqBEdwHDsHKHtdp4=', 'last_login': None, 'is_superuser': False, 'username': 'myuser', 'first_name': '', 'last_name': '', 'email': '', 'is_staff': False, 'is_active': True, 'date_joined': datetime.datetime(2024, 7, 18, 8, 51, 50, 785542, tzinfo=datetime.timezone.utc), 'group_names': None}, {'id': 1, 'password': 'pbkdf2_sha256$600000$ZvhMM2FoA2ecuJNcy9Axg5$FExRSNq/cKmRQO7v1aoa25Jbsz1R1M0moCdEyG03DzE=', 'last_login': datetime.datetime(2024, 7, 29, 8, 17, 31, 660706, tzinfo=datetime.timezone.utc), 'is_superuser': True, 'username': 'admin', 'first_name': '', 'last_name': '', 'email': 'admin@admin.com', 'is_staff': True, 'is_active': True, 'date_joined': datetime.datetime(2024, 7, 18, 8, 51, 26, 777004, tzinfo=datetime.timezone.utc), 'group_names': None}]>

<QuerySet [{'id': 1, 'name': 'Group'}, {'id': 2, 'name': 'Group 1'}]>

```

`after_export ` does not seem to take into account the overriding of the `get_queryset` method.

sergei-maertens commented on auth_matrix/admin_views.py:
```diff
@@ -1,15 +1,27 @@
+        return self.request.user.is_staff
+
+
+def get_auth_matrix_permission():
+    if User._meta.app_label == "auth":
+        return "auth.view_user"
+    return "accounts.view_user"
```
the pattern is `<app_label>.view_<model_name>`, so you can build this permission name exactly without guessing:

```py
opts = User._meta
permission = f"{opts.app_label}.view_{opts.model_name}"
```

sergei-maertens commented on auth_matrix/templates/admin/accounts/user/change_list.html:
```diff
@@ -1,27 +1,8 @@
-
-{% block pagination %}
+{% block object-tools-items %}
+    {% include "admin/auth_matrix/includes/object_tools_items.html" %}
```
proper namespacing would swap the directory structure: `auth_matrix/admin/includes/object_tools_items.html`

sergei-maertens commented on auth_matrix/templates/admin/accounts/user/change_list.html:
I don't think you should include this file at all - the `accounts` app is something that is part of our default-project, but including the template here removes the ability of other apps/layers to override the templates.

I think shipping the include template is enough, and in the `testapp/` you can specify a different template to implement your tests.

Documentation will then need to include the steps to include/add the snippet in your own project templates, of course.

sergei-maertens commented on auth_matrix/templates/admin/auth/group/change_list.html:
Same remarks as in `auth_matrix/templates/admin/accounts/user/change_list.html`

sergei-maertens commented on auth_matrix/templates/admin/auth_matrix/includes/object_tools_items.html:
```diff
@@ -0,0 +1,10 @@
+{% load i18n %}
+{% if request.user.is_staff %}
+    {% if perms.auth.view_user or perms.accounts.view_user %}
```
* I think the staff check is not needed, since we're doing a permission check already
* we'll need a different solution for the permission check based on what the actual user model is, see the remark in the python code

sergei-maertens commented on pyproject.toml:
```diff
@@ -27,14 +27,17 @@ classifiers = [
-    "django>=4.2"
+    "django>=4.2",
+    "django-import-export",
+    "psycopg2",
```
this breaks our ability to use `psycopg` instead of `psycopg2` - better to just document that PostgreSQL is required and not specify this dependency.

sam-bertin replied:
right but this won't build now. It needs it here: https://github.com/maykinmedia/maykin-auth-matrix/actions/runs/10145474467/job/28051237009?pr=1 so if we can't do that I did not find a working alternative.

sam-bertin replied:
are you suggesting we should remove all psycopg deps? 

How is the CI able to pass without having psycopg to dialog with the DB ?

sergei-maertens replied:
the easiest way is to add it to the `tests` extras dependency group: https://github.com/maykinmedia/maykin-auth-matrix/blob/main/pyproject.toml#L40

sergei-maertens commented on pyproject.toml:
```diff
@@ -27,14 +27,17 @@ classifiers = [
+    "django>=4.2",
+    "django-import-export",
+    "psycopg2",
+    "python-dotenv",
```
why is this needed?

sergei-maertens commented on testapp/settings.py:
```diff
@@ -1,5 +1,12 @@
+import os
 from pathlib import Path
 
+from dotenv import load_dotenv
+
+load_dotenv()  # take environment variables from .env.
```
IMO this is too complicated for a testapp - instead we just set the envvars directly in the `ci.yml` workflow file and you can easily set them in your virtualenv for local development. See https://github.com/maykinmedia/zgw-consumers/blob/main/testapp/settings.py#L10

sergei-maertens commented on docs/quickstart.rst:
```diff
@@ -2,17 +2,84 @@
-    pip install auth_matrix
+      from auth_matrix.views import AuthMatrixView
 
+      urlpatterns = [
+          path(
+              "admin/auth_matrix/",
+              include(("auth_matrix.admin_urls", "auth_matrix"), namespace="auth_matrix"),
+          ),
+      ]
```
setting `app_name = "auth_matrix"` in the URLs file allows you to simplify this to:

```suggestion
      urlpatterns = [
          path("admin/auth_matrix/", include("auth_matrix.admin_urls")),
      ]
```

sam-bertin committed:
[#399] Processed Review

- 🔥Removed unused files
- 🎨 Make use of mixins to hook up modelAdmin instead of hard coding the admin into the app
- ♻️ Proper Namespacing templates structure.
- ➖Removed unecessary deps
- 🔧 Simplified env conf
- 🛂 Custom permissions
- 📝 Updated documentation

sergei-maertens commented on .gitignore:
```diff
@@ -101,4 +103,4 @@ venv.bak/
 
 # mypy
-.mypy_cache/
+.mypy_cache/
```
please don't remove the final newline in a file

it seems like your editor is not picking up the `.editorconfig` file (see rule [here](https://github.com/maykinmedia/maykin-auth-matrix/blob/main/.editorconfig#L7))

sergei-maertens commented on auth_matrix/admin.py:
```diff
@@ -32,22 +36,23 @@ class Meta:
-        return dataset
+
+class UserExportMixin(ExportMixin):
+    resource_classes = (UserGroupResource,)
```
I keep getting confused by the name of this resource, so perhaps renaming it to `UserWithGroupsResource` will make it more obvious that we are indeed dealing with a user resource (and then the queryset/model confusion from earlier should hopefully not happen again).

sergei-maertens commented on auth_matrix/admin.py:
```diff
@@ -32,22 +36,23 @@ class Meta:
+
+class UserExportMixin(ExportMixin):
+    resource_classes = (UserGroupResource,)
+    change_list_template = "admin/auth/user/change_list.html"
```
IMO the library should not specify what the `change_list_template` is, because this conflicts if other libraries do this.

django-import-export even has an FAQ entry about this: https://django-import-export.readthedocs.io/en/latest/faq.html#failed-to-assign-change-list-template-attribute-warning-in-logs

Proper design is that the library offers the necessary template fragments that the project making use of the library can (and should) then `{% include ... %}` themselves, and that should be part of the installation instructions. So those installation instructions need a code example of how the admin template should be specified and which snippet needs to be included.

sam-bertin replied:
so you think I should remove it and include this ?

```rst
8. Include the necessary template fragments in your admin templates to enable the authorization matrix functionality. For example, if you override the default User or Group model, you can add this template to your template model's folder (e.g. ``templates/admin/auth/group/change_list.html``):

   .. code-block:: django
      {% extends "admin/change_list.html" %}

      {% load admin_list i18n %}

      {% block object-tools-items %}
          {% include "auth_matrix/admin/includes/object_tools_items.html" %}
          {{ block.super }}
      {% endblock %}

9. Config the template location in the mixin with the matching template.

    .. code-block:: python 
    
       class GroupExportMixin(ExportMixin):
           resource_classes = (GroupPermissionResource,)
           change_list_template = "admin/auth/group/change_list.html"
```

sam-bertin replied:
How about this ? https://github.com/django-import-export/django-import-export/issues/1514#issuecomment-1344200867

sam-bertin replied:
Nvm, it makes the export button disappear this way. I can't find a way to re include it. Tried to include the default template location (admin/import_export/change_list_export.html)   within the custom template.

sergei-maertens commented on auth_matrix/admin.py:
```diff
@@ -61,38 +66,24 @@ class Meta:
-    def dehydrate(self, group):
-        data = super().dehydrate(group)
+    def after_export(self, queryset, data, *args, **kwargs):
         permission_names = [permission.name for permission in Permission.objects.all()]
```
```suggestion
        permission_names = Permission.objects.values_list("name", flat=True)
```
you can do this in a single optimized query :)

sergei-maertens commented on auth_matrix/admin.py:
```diff
@@ -61,38 +66,24 @@ class Meta:
-    def export(self, queryset, *args, **kwargs):
-        dataset = super().export(queryset, *args, **kwargs)
-        permission_names = [permission.name for permission in Permission.objects.all()]
-        for permission_name in permission_names:
-            dataset.append_col(
+            data.append_col(
                 [
                     group.permissions.filter(name=permission_name).exists()
                     for group in queryset
                 ],
                 header=permission_name,
             )
```
@alextreme I see that the permission `name` field is an English non-unique string created by django when the permission is created. I think it'd be better to use the `code_name` and `content_type` field together to report the permission, or at least use `str(permission)` which includes the content type?

In its current form, you could have the same/similar model name in different Django apps (as we do with `Log` from yubin/django-timeline-logger/admin log/...) and the permissions for a given group could be reported incorrectly, because the content-type is not taken into account.

This could yield a permission as `myapp_MyModel:can_view_mymodel`, so, I suppose you could clean up this string to `myapp:can_view_mymodel`. On the other hand, when I define custom permission `foo` on two different models in the same app, they are different permission records and they would both show as `myapp:foo`, so maybe `myapp:mymodel:foo` would be best to make it non-ambiguous

sergei-maertens commented on auth_matrix/admin.py:
```diff
@@ -61,38 +66,24 @@ class Meta:
-    filter_horizontal = ("permissions",)
+class GroupExportMixin(ExportMixin):
+    resource_classes = (GroupPermissionResource,)
     change_list_template = "admin/auth/group/change_list.html"
```
similar remarks as with the `UserExportMixin` about the template overriding

sergei-maertens commented on auth_matrix/admin_urls.py:
```diff
@@ -1,10 +1,14 @@
+app_name = "auth_matrix"
+
 urlpatterns = [
     path(
         "matrix",
-        AuthorizationMatrixView.as_view(),
+        admin.site.admin_view(AuthorizationMatrixView.as_view()),
         name="authorization_matrix",
     ),
 ]
```
IMO this should be part of the user admin mixin and override the `ModelAdmin.get_urls` method to add additional URLs, instead of having to hook this URL up manually in your urlconf.

sergei-maertens commented on auth_matrix/admin_views.py:
```diff
@@ -1,15 +1,37 @@
+    return f"{opts.app_label}.view_{opts.model_name}"
+
+
+class CanViewAuthorizationMatrixMixin(UserIsStaffMixin):
+    def dispatch(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
+        if not request.user.has_perm(get_auth_matrix_permission_label()):
+            return self.handle_no_permission()
+        return super().dispatch(request, *args, **kwargs)
```
The `PermissionRequiredMixin` already has a hook to dynamically return/calculate the required permission: `get_permission_required`

```py
class CanViewAuthorizationMatrixMixin(UserIsStaffMixin, PermissionRequiredMixin):

    def get_permission_required(self):
        permission_required = get_auth_matrix_permission_label()
        return (permission_required,)
```

this has a couple of benefits:

1. avoiding overriding `dispatch` (since mistakes in here have high risk, security wise)
2. the order of checks: first `UserIsStaffMixin` is evaluated, and it shortcuts if the user is not logged in or staff. Next, the built-in `PermissionRequiredMixin` is checked, and we validate it against the user model permission (dynamically determined)
3. No need to override any "unhappy flow" handling

All of the above runs during the `dispatch` phase of the view, and it's all going via inheritance. Only when all the requirements are met, the actual view is executed.

sergei-maertens commented on auth_matrix/admin_views.py:
```diff
@@ -1,15 +1,37 @@
+
+class AuthorizationMatrixView(
+    CanViewAuthorizationMatrixMixin,
+    views.generic.base.TemplateView,
```
please just use `from django.views.generic import TemplateView`

sergei-maertens commented on auth_matrix/admin_views.py:
```diff
@@ -1,15 +1,37 @@
+    template_name = "auth_matrix/admin/authorization_matrix.html"
 
     def get(self, request):
         groups = Group.objects.prefetch_related("permissions").order_by("name")
```
I think this has the same issues as the export w/r to the permission names not being unique and filtering on name especially may show an auth matrix that is wrong.

It's a good exercise for you to write a single `Group` or `Permission` query that already includes the necessary information via `prefetch` so that you can avoid these problems :)

sam-bertin replied:
Here we are ordering groups by name so that should not be an issue since they are unique.
But I understand the concern for permissions. I have submitted the changes.

sergei-maertens replied:
right I had some PR review blindness here, it's already using the prefetch so you can disregard my remark. It's hard to review code that's building on top of unreviewed code.

sergei-maertens commented on auth_matrix/templates/auth_matrix/admin/authorization_matrix.html:
```diff
@@ -11,7 +11,8 @@
     .matrix-table td {
     border: 1px solid black;
-    padding: 8px;
+    padding: 10px;
+    font-size: 14px;
```
so the purpose of inheriting django templates was that we ideally could get rid of all custom CSS - I see that you haven't succeeded with that yet. Shall we spend some time this week to figure this out?

sam-bertin replied:
yeah :+1: if you find time

sergei-maertens commented on auth_matrix/templates/auth_matrix/admin/includes/object_tools_items.html:
```diff
@@ -0,0 +1,8 @@
+{% load i18n %}
+{% if can_view_auth_matrix %}
+    <li>
+        <a href="{% url 'auth_matrix:authorization_matrix' %}" class="button mt">
```
one advantage of adding the URLs via the admin mixin, is that we will actually be able to use a URL name that doesn't have namespace issues :)

sam-bertin replied:
I'm not sure that I have understood what you meant here.
Isn't this already making use of the changes we made ?

sergei-maertens replied:
django has some mechanisms making it possible to load an app multiple times with different namespaces. however, reversing URLs requires you to know all the namespaces upfront, which is of course not possible in a library. users may include the url patterns in different ways, but that breaks the app because then it cannot reverse the URL name anymore

All of these problems should go away automagically when this url/view is part of the admin views through the `admin.ModelAdmin.get_urls` method, since it takes care of properly resolving the app and we don't have to worry about it. So then you'll use `admin:user_authorization_matrix` for example (if that's the URL name that we'll go with).

sam-bertin replied:
I don't get this. The authorization matrix works for both groups and user. Attaching the url to both model wouldnt allow me to use the same url pattern on both models. Having different urls would't allow me to include the `object_tools_items.html ` template. I don't think I understand the suggestion.

sam-bertin replied:
Eventually, this seems wrong to me but this work :

```
class DummyModel(models.Model):
    pass

    class Meta:
        managed = False


class DummyModelAdmin(admin.ModelAdmin):
    pass

    def get_urls(self):
        from django.urls import path

        return [
            path(
                "matrix/",
                self.admin_site.admin_view(AuthorizationMatrixView.as_view()),
                name="authorization_matrix",
            )
        ]


admin.site.register(DummyModel, DummyModelAdmin)
```

sam-bertin replied:
ok - discussed together

sergei-maertens reviewed (commented):
There is more to review, but let's focus on these aspects first otherwise things will get too overwhelming!

sam-bertin committed:
[#399] Processed review

⚡️ Query optimization
🎨 Include PermissionMixin instead of dispatch method

sam-bertin committed:
[#399] 🏗️ Moved templates to the testapp folder

Remove templates from the delivered app and add the snippet instructions to the docs so that the user can set it up himself.

sam-bertin committed:
[#399] 🚸 Improve usability

Make use of the get_urls instead of asking the user to set up the urls himself.

sam-bertin committed:
Revert "[#399] 🚸 Improve usability"

This reverts commit 67faa7d16002fd53d1946f6aa6450b64dcc6f7bc.

sam-bertin committed:
Revert "[#399] 🏗️ Moved templates to the testapp folder"

This reverts commit 137895ae734007b036d1e5de6157838feb2c03c6.

sam-bertin committed:
[#399] 🏗️ Moved templates to the testapp folder

Remove templates from the delivered app and add the snippet instructions to the docs so that the user can set it up himself.

sam-bertin committed:
[#399] Update urls

Processing review

sam-bertin committed:
🚸 Improve usability

Make use of the get_urls instead of asking the user to set up the urls himself.

sam-bertin committed:
🚨 Linting / Formatting

sam-bertin committed:
📝 Template instructions

Added instructions for user to configure their template on the mixin if they ever happen to have different ones. This is part of minimizing the user actions when using our library.

sam-bertin committed:
💄 Added custom breadcrumbs

sam-bertin committed:
💄Updated table CSS

Updated the custom CSS in my matrix template to make use of the Django default admin CSS classes. Still some custom CSS left to make it more presentable.

sam-bertin commented:
@sergei-maertens I tried to addressed the issues mentioned. Maybe we can go for round 2.
